Docs: map task (#209)
latchbio/latch: docs/source/basics/map_task.md

# Map Task

There are many pipelines in bioinformatics that require running a processing
step in parallel and aggregating their outputs at the end for downstream

> analysis. A prominent example of this is bulk RNA-sequencing, where the

Reviewer: ```suggestion
analysis. A prominent example of this is bulk RNA-sequencing, where
```

alignment is performed to produce transcript abundances per sample, and gene
counts of all samples are subsequently merged. Having a single count matrix
makes it convenient to use in downstream steps, such as differential gene
expression analysis. Another example is performing FastQC on multiple samples
and summarizing the results in a MultiQC report.

The Latch SDK introduces a construct called `map_task` to help parallelize a

> task across a list of inputs. This means you can run thousands of instances of

Reviewer: ```suggestion
task across a list of inputs. This means you can run multiple instances of
```

id rather not make unrealistic claims

the task at the same time inside a single workflow, providing valuable
performance gains.

Let's look at a simple example below!

First, import `map_task` into your workflow:

```python
from typing import List

from latch import map_task, small_task, workflow
```

Next, define a task to use in the map task.
> Note: A map task can only accept **one input** and produce **one output**.

```python
@small_task
def a_mappable_task(a: int) -> str:
    inc = a + 2
    stringified = str(inc)
    return stringified
```

Let's also define a task that collects the mapped output and returns a string:

```python
@small_task

> def coalesce(b: typing.List[str]) -> str:

Reviewer: ```suggestion
def coalesce(b: List[str]) -> str:
```

you already imported `List` from `typing` above

    coalesced = "".join(b)
    return coalesced
```

> We send `a_mappable_task` to be repeated across a collection of inputs to the
> `map_task()` function. The task `a_mappable_task` is run for each element in
> the list.

Reviewer: ```suggestion
We can run `a_mappable_task` across a collection of inputs using the `map_task` function. This function takes in `a_mappable_task` and returns a mapped version of that task. This mapped version takes as input a list of inputs to `a_mappable_task` , and returns a list of the outputs of `a_mappable_task` run on all inputs in the list in parallel.
```

```python
@workflow
def my_map_workflow(a: typing.List[int]) -> str:
    mapped_out = map_task(a_mappable_task)(a=a)
    coalesced = coalesce(b=mapped_out)
    return coalesced
```

That's it! You've successfully defined `a_mappable_task` that is passed to a

> `map_task()` and ran repeatedly on a list of inputs in parallel. You have also

Reviewer: ```suggestion
`map_task()` and run repeatedly on a list of inputs in parallel. You have also
```

> defined a `coalesce` task to collect the list of outputs from the mapped task

Reviewer: ```suggestion
defined a `coalesce` task that collects the list of outputs from the mapped task
```

and returns a string.

## Map a Task with Multiple Inputs

> You may want to map a task with multiple inputs.

Reviewer: ```suggestion
Often, you want to map a task with multiple inputs.
```

For example, the task below takes in 2 inputs, a base and a DNA sequence, and

> returns percentage of that base in the sequence:

Reviewer: ```suggestion
returns the percentage of that base in the sequence:
```

```

Let's use the new `mappable_task` in our workflow:

```python
@workflow
def count_wf(list_base: List[str] = ["A", "T", "C", "G"], dna_sequence: str = "AAAATTTCCGG") -> List[float]:
    prepared = prepare_map_inputs(list_base=list_base, dna_sequence=dna_sequence)
    return map_task(mappable_task)(input=prepared)
```

Great! Now, we are able to use the `count_wf` to spin up four tasks in
parallel. The `map_task` returns a list of four floats, each of which is the
percentage of base pair in the DNA sequence.

---

## Bonus: Learning through a Biological Example

In the example below, we walk through a practical example of how we can use the
map task construct to run FastQC on multiple samples and summarize their
results in a MultiQC report.

First, we define a Dataclass that contains a sample name and its associated
FastQ file:

```python
@dataclass_json
@dataclass
class Sample:
    sample_name: str
    fastq: LatchFile
```

Then, we create a task to run FastQC on a single sample and output the result
under the **FastQC Results** folder on Latch.

```python
@small_task
def fastqc_task(sample) -> LatchDir:

>     outdir = "/root/fastqc_result"
>     os.mkdir(outdir)

Reviewer: ```suggestion
    outdir = Path("/root/fastqc_result").resolve()
    outdir.mkdir(exist_ok=True)
```

please use pathlib

    _fastqc_cmd = [
        "/root/FastQC/fastqc", 
        sample.fastq.local_path, 
        f"--outdir={outdir}"
    ]

>     subprocess.run(_fastqc_cmd)

Reviewer: ```suggestion
    subprocess.run(_fastqc_cmd, check=True)
```

    return LatchDir("/root/fastqc_result", f"latch:///FastQC Results/{sample.sample_name}")
```

> **Concept check**: Note how this task will later be mapped across a list of
samples. Therefore, the task is defined to accept one input and return one
output.

Next, define a second task to run MultiQC on a given directory for analysis
logs and compiles a HTML report.

```python
@small_task
def multiqc_task(fastqc_results: List[LatchDir]) -> LatchDir:

    outdir = "/root/multiqc_results"

>     os.mkdir(outdir)

Reviewer: same as above, pathlib over os pls

    fastqc_dirs = [result.local_path for result in fastqc_results]

    _multiqc_cmd = ["multiqc"] + fastqc_dirs + ["-o", outdir]

>     subprocess.run(_multiqc_cmd)

Reviewer: check=True here too
